File: my-solutions/0039-combination-sum/solution.cpp
class Solution {
public:
 
    void findUnique(vector<vector<int>>& ans, vector<int>& temp, vector<int>& candidates,int target,int i){
        if(i == candidates.size()){
            if(target == 0){
                ans.push_back(temp);
            }
            return;
        }

        if(candidates[i] <= target){
            temp.push_back(candidates[i]);
            findUnique(ans,temp,candidates,target-candidates[i],i);
            temp.pop_back();
        }
        findUnique(ans,temp,candidates,target,i+1);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>temp;
        findUnique(ans,temp,candidates,target,0);
        return ans;
    }
};

File: my-solutions/0039-combination-sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort(reverse=True)
        result = []

        for i in range(len(candidates)):
            if candidates[i] > target:
                continue
            result += self.helper(candidates, [candidates[i]], candidates[i], target, i)
        
        return result
    

    def helper(self, candidates: List[int], running: List[int], total: int, target: int, index: int) -> List[List[int]]:
        if total == target:
            return [running]
        
        solution = []
        for i in range(index, len(candidates)):
            if total + candidates[i] > target:
                continue
            solution += self.helper(candidates, running + [candidates[i]], total + candidates[i], target, i)
        
        return solution
